doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,291 @@
+
+
+### Security Checks
+
+OpenSSL currently defines configurable FIPS options.
+These options are supplied via the FIPS configuration file - which is normally setup via fipsinstall.
+
+- FIPS_FEATURE_CHECK(FIPS_security_check_enabled, fips_security_checks, 1)
+- FIPS_FEATURE_CHECK(FIPS_tls_prf_ems_check, fips_tls1_prf_ems_check, 0)
+- FIPS_FEATURE_CHECK(FIPS_restricted_drbg_digests_enabled, 0)
+- OSSL_PROV_FIPS_PARAM_CONDITIONAL_ERRORS selftest_params.conditional_error_check
+
+The following functions are available in providers/common/security_check.c.
+
+- ossl_rsa_check_key()
+- ossl_ec_check_key()
+- ossl_dsa_check_key()
+- ossl_dh_check_key()
+- ossl_digest_get_approved_nid_with_sha1()
+- ossl_digest_is_allowed()
+
+Anywhere where these functions are called an indicator MAY be required.
+Because these options are available I do not think it is sufficient to
+document this in the security policy.
+
+Each of these functions contains code of the following form:
+~~~
+#if !defined(OPENSSL_NO_FIPS_SECURITYCHECKS)
+    if (ossl_securitycheck_enabled(ctx)) {
+      // Do some checks, and maybe return 0 for a hard failure
+      ...
+    }
+}
+~~~
+
+OPENSSL_NO_FIPS_SECURITYCHECKS is also a configurable option
+If the security checks are not enabled then it is unapproved?
+
+
+## Implementation options
+
+The above requirements indicate 2 options.
+
+### Option 1
+- Dont allow ANY non approved algorithms and then a indicator is not required.
+
+#### Pros
+- Simple
+
+#### Cons
+- Problematic since we already have configurable options that are used for security checks etc.
+- We would need to return errors anywhere where an algorithm is not approved, which would cause
+  compatability issues
+
+### Preferred Option
+- Add an indicator everywhere that it is required.
+
+#### Pros
+- Flexible solution
+#### Cons
+- Requires a lot more effort to add the indicator to all the required places.
+
+
+Note that in order for a service to be ‘fips approved’ the following requirments would need to be met.
+- Any algorithms are fetched using “fips=yes”
+- A service is a series of one or more API calls that must all succeed
+- A extra API call is needed after the service succeeds, that should return 1 if the service is approved.
+
+#### Solution 1 (Using an indicator everywhere)
+
+Use a per thread global counter that is incremented when an algorithm is approved.  AWS/google have added this in places where a service is at the point of completing its output (e.g. digest_final). This design is complicated by the fact that a service may call another service (e.g HMAC using SHA256) that also wants to increment the approved counter. To get around this issue they have a second variable that is used for nesting. If the variable is non zero then the approved counter doesnt increment. This also allows non security relevant functions to not increment the approved count. Another variation of this would be to use flags instead of a counter.
```

Comment:
IMO: this is rubbish.

Reply:
It is a different way of doing it, I have listed why this is not a good idea.